platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/orphan/OrphanBuffer.java

```diff
@@ -0,0 +1,254 @@
+     *
+     * @param minimumGenerationNonAncient the minimum generation of non-ancient events to keep in the buffer
+     */
+    public void setMinimumGenerationNonAncient(final long minimumGenerationNonAncient) {
+        this.minimumGenerationNonAncient = minimumGenerationNonAncient;
+
+        eventsWithParents.shiftWindow(minimumGenerationNonAncient);
+
+        // As the map is cleared out, we need to gather the ancient parents and their orphans. We can't
+        // modify the data structure as the window is being shifted, so we collect that data and act on
+        // it once the window has finished shifting.
+        final List<ParentAndOrphans> ancientParents = new ArrayList<>();
+        missingParentMap.shiftWindow(
+                minimumGenerationNonAncient,
+                (parent, orphans) -> ancientParents.add(new ParentAndOrphans(parent, orphans)));
+
+        ancientParents.forEach(this::missingParentBecameAncient);
+    }
+
+    /**
+     * Called when a parent becomes ancient.
+     * <p>
+     * Accounts for events potentially becoming un-orphaned as a result of the parent becoming ancient.
+     *
+     * @param parentAndOrphans the parent that became ancient, along with its orphans
+     */
+    private void missingParentBecameAncient(@NonNull final ParentAndOrphans parentAndOrphans) {
+        final EventDescriptor parentDescriptor = parentAndOrphans.parent();
+
+        for (final OrphanedEvent orphan : parentAndOrphans.orphans()) {
+            orphan.missingParents().remove(parentDescriptor);
+
+            if (orphan.missingParents().isEmpty()) {
+                eventIsNotAnOrphan(orphan.orphan());
+            }
+        }
+    }
+
+    /**
+     * Get the parents of an event that are currently missing.
+     *
+     * @param event the event whose missing parents to find
+     * @return the list of missing parents, empty if no parents are missing
+     */
+    @NonNull
+    private List<EventDescriptor> getMissingParents(@NonNull final GossipEvent event) {
+        final List<EventDescriptor> missingParents = new ArrayList<>();
+
+        final Iterator<EventDescriptor> parentIterator = new ParentIterator(event);
+        while (parentIterator.hasNext()) {
+            final EventDescriptor parent = parentIterator.next();
+            if (!eventsWithParents.contains(parent) && parent.getGeneration() >= minimumGenerationNonAncient) {
+                missingParents.add(parent);
+            }
+        }
+
+        return missingParents;
+    }
+
+    /**
+     * Signal that an event is not an orphan.
+     * <p>
+     * Accounts for events potentially becoming un-orphaned as a result of this event not being an orphan.
+     *
+     * @param event the event that is not an orphan
+     */
+    private void eventIsNotAnOrphan(@NonNull final GossipEvent event) {
+        final Deque<GossipEvent> nonOrphanStack = new LinkedList<>();
+        nonOrphanStack.push(event);
+
+        while (!nonOrphanStack.isEmpty()) {
```

Comment:
I suggest adding a comment something like this for future engineers who might lack some context:

```
// When a missing parent is found, there may be many descendants of that parent who end up
// being un-orphaned. This loop frees all such orphans non-recursively (recursion yields pretty
// code but can thrash the stack).
```